encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java

```diff
@@ -0,0 +1,185 @@
+package openlr.encoder.routesearch;
+
+import openlr.encoder.OpenLREncoderProcessingException;
+import openlr.map.GeoCoordinates;
+import openlr.map.Line;
+import openlr.map.utils.GeometryUtils;
+import openlr.map.utils.PQElem;
+
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+/**
+ * <h1>Verifying the existence of Fully or partially joined second shortest route with length less than relative threshold</h1>
+ */
+public class SecondShortestRouteChecker {
+    private GeoCoordinates destination;
+    private List<? extends Line> location;
+    private int maxLengthAllowed;
+    private BiFunction<Integer, Integer, Boolean> verifyThreshold;
+
+
+    private SecondShortestRouteChecker(GeoCoordinates destination, List<? extends Line> location, int maxLengthAllowed, BiFunction<Integer, Integer, Boolean> verifyThreshold) {
+        this.destination = destination;
+        this.location = location;
+        this.maxLengthAllowed = maxLengthAllowed;
+        this.verifyThreshold = verifyThreshold;
+    }
+
+
+    /**
+     *
+     * @param location route
+     * @param relativeThreshold threshold relative to the route length
+     * @return Instance of SecondShortestRouteChecker for the location with threshold
+     * @throws OpenLREncoderProcessingException
+     */
+    public static SecondShortestRouteChecker on(List<? extends Line> location, double relativeThreshold) throws OpenLREncoderProcessingException {
+        if (location.isEmpty()) {
+            throw new OpenLREncoderProcessingException(OpenLREncoderProcessingException.EncoderProcessingError.NO_ROUTE_FOUND_ERROR);
+        }
+        GeoCoordinates destinationStart = location.get(location.size() - 1).getStartNode().getGeoCoordinates();
+        int locationLength = location.stream().mapToInt(Line::getLineLength).sum();
+
+        int maxLengthAllowed = locationLength + (int) (locationLength * relativeThreshold);
+
+        BiFunction<Integer, Integer, Boolean> verifyThreshold = (Integer lengthAlongSecondShortestRoute, Integer lengthAlongLocation) -> {
```

Comment:
You could just make this function an instance method and make the `relativeThreshold` a member variable.  I think defining it as a `BiFunction` member variable is a bit unusual and requires a bit more mental parsing :-)

```
private boolean verifyThreshold(int lengthAlongSecondShortestRoute, int lengthAlongLocation) {
    return lengthAlongSecondShortestRoute < (lengthAlongLocation + (int) (lengthAlongLocation * this.relativeThreshold);
}
```

Reply:
d14a5b816e9b91dd99d4db70246353f967710764